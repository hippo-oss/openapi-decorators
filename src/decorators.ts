import { DTODecoratorFactories } from '@hippo-oss/dto-decorators';

const noop = () => null;
const noopFactory = () => noop;

export const OPENAPI_DECORATORS: DTODecoratorFactories = {
    IsBoolean: noopFactory,
    IsDate: noopFactory,
    IsDateString: noopFactory,
    IsEnum: noopFactory,
    IsInteger: noopFactory,
    IsNested: noopFactory,
    IsNumber: noopFactory,
    IsString: noopFactory,
    IsUUID: noopFactory,
};

export const {
    IsBoolean,
    IsDate,
    IsDateString,
    IsEnum,
    IsInteger,
    IsNested,
    IsNumber,
    IsString,
    IsUUID,
} = OPENAPI_DECORATORS;
