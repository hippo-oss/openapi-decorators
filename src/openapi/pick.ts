import 'reflect-metadata';
import { Constructor, Target } from '@hippo-oss/dto-decorators';
import { trimStart } from 'lodash';

/* @nestjs/swagger does not expose its internal metadata operations, so we partially
 * reproduce them here. Obviously, this code takes advantage of internal functionality
 * and may be subject to regression if the underlying library changes its implementation.
 */

/* The Reflect metadata key used to store the array of OpenAPI properties.
 */
const OPENAPI_PROPERTIES_ARRAY = 'swagger/apiModelPropertiesArray';

/* Set the the OpenAPI properties array.
 */
export function setPropertiesArray<T, F extends keyof T>(cls: Constructor<T>, fields: F[]): void {
    const propertiesArray = fields.map(
        (name: F): string => `:${String(name)}`,
    );

    // eslint-disable-next-line @typescript-eslint/ban-types
    Reflect.defineMetadata(OPENAPI_PROPERTIES_ARRAY, propertiesArray, cls.prototype as {});
}

/* Get the the OpenAPI properties array as fields.
 */
export function getPropertiesArray<T, F extends keyof T>(cls: Constructor<T>): F[] {
    // eslint-disable-next-line @typescript-eslint/ban-types
    const propertiesArray = Reflect.getMetadata(OPENAPI_PROPERTIES_ARRAY, cls.prototype as {}) as unknown as string[];

    return (propertiesArray || []).map(
        (name: string) => trimStart(name, ':') as F,
    );
}

/* Narrow OpenAPI properties by picking the provided fields.
 */
export function pickOpenAPIProperties<T extends Target, F extends keyof T, I extends keyof T>(
    cls: Constructor<T>,
    fields: F[],
    ignoreFields: I[] = [],
): Constructor<T> {
    const propertiesArray: (keyof T)[] = getPropertiesArray(cls);

    setPropertiesArray(
        cls,
        propertiesArray.filter(
            (field) => fields.includes(field as F) && !ignoreFields.includes(field as I),
        ),
    );

    return cls;
}
