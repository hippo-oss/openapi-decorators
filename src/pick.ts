import { Constructor, Target } from '@hippo-oss/dto-decorators';

import { pickOpenAPIProperties } from './openapi';

/* Create a derived DTO class from the `Base` DTO class.
 */
export function derive<T extends Target>(
    Base: Constructor<T>,
    name: string,
): Constructor<T> {
    // NB: anonymous classes are named via their assigned variable
    return {
        [name]: class extends Base { },
    }[name];
}

/* Create a derived class that narrows the `Base` class by picking fields.
 */
export function pick<T extends Target, F extends keyof T, I extends keyof T>(
    Base: Constructor<T>,
    name: string,
    fields: F[],
    ignoreFields: I[] = [],
): Constructor<Pick<T, F>> {

    const operators = [
        pickOpenAPIProperties,
    ];

    return operators.reduce(
        (cls, operator) => operator(cls, fields, ignoreFields),
        derive(Base, name),
    );
}
