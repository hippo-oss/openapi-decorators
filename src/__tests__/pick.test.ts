import { get } from 'lodash';

import { IsNested, IsInteger, IsString } from '../decorators';
import { pick } from '../pick';
import { createController, createDocument } from './fixtures';

type Obj = Record<string, unknown>;

describe('pick', () => {
    class Child {
        @IsString()
        baz!: string;

        @IsString()
        qux!: string;
    }

    class Parent {
        @IsString()
        foo!: string;

        @IsString()
        bar!: string;

        @IsNested({
            type: Child,
        })
        child!: Child;
    }

    it('picks a single field', async () => {
        const ParentFixture = pick(Parent, 'ParentFixture', ['foo']);

        const Controller = createController(ParentFixture);

        const document = await createDocument(Controller);
        const response = document.paths['/example']?.get?.responses[200];
        expect(response).toMatchObject({
            content: {
                'application/json': {
                    schema: {
                        $ref: '#/components/schemas/ParentFixture',
                    },
                },
            },
        });

        const properties = get(document, 'components.schemas.ParentFixture.properties') as unknown as Obj;
        const required = get(document, 'components.schemas.ParentFixture.required') as unknown[];
        expect(Object.keys(properties)).toEqual(['foo']);
        expect(required).toEqual(['foo']);

        expect(document).toMatchSnapshot();
    });
    it('picks zero fields', async () => {
        const ParentFixture = pick(Parent, 'ParentFixture', []);

        const Controller = createController(ParentFixture);

        const document = await createDocument(Controller);
        const response = document.paths['/example']?.get?.responses[200];
        expect(response).toMatchObject({
            content: {
                'application/json': {
                    schema: {
                        $ref: '#/components/schemas/ParentFixture',
                    },
                },
            },
        });

        const properties = get(document, 'components.schemas.ParentFixture.properties') as unknown as Obj;
        const required = get(document, 'components.schemas.ParentFixture.required') as unknown[];
        expect(Object.keys(properties)).toEqual([]);
        expect(required).not.toBeDefined();

        expect(document).toMatchSnapshot();
    });
    it('picks child fields', async () => {
        const ParentFixture = pick(Parent, 'ParentFixture', ['child']);

        const Controller = createController(ParentFixture);

        const document = await createDocument(Controller);
        const response = document.paths['/example']?.get?.responses[200];
        expect(response).toMatchObject({
            content: {
                'application/json': {
                    schema: {
                        $ref: '#/components/schemas/ParentFixture',
                    },
                },
            },
        });

        const parentProperties = get(document, 'components.schemas.ParentFixture.properties') as unknown as Obj;
        const parentRequired = get(document, 'components.schemas.ParentFixture.required') as unknown[];
        expect(Object.keys(parentProperties)).toEqual(['child']);
        expect(parentRequired).toEqual(['child']);

        const childProperties = get(document, 'components.schemas.Child.properties') as unknown as Obj;
        const childRequired = get(document, 'components.schemas.Child.required') as unknown[];
        expect(Object.keys(childProperties)).toEqual(['baz', 'qux']);
        expect(childRequired).toEqual(['baz', 'qux']);

        expect(document).toMatchSnapshot();
    });
    it('picks one field and one child field', async () => {
        const ChildFixture = pick(Child, 'ChildFixture', ['baz']);
        const ParentBase = pick(Parent, 'ParentBase', ['foo'], ['child']);

        class ParentFixture extends ParentBase {
            @IsNested({
                type: ChildFixture,
            })
            child!: typeof ChildFixture;
        }

        const Controller = createController(ParentFixture);

        const document = await createDocument(Controller);
        const response = document.paths['/example']?.get?.responses[200];
        expect(response).toMatchObject({
            content: {
                'application/json': {
                    schema: {
                        $ref: '#/components/schemas/ParentFixture',
                    },
                },
            },
        });

        const parentProperties = get(document, 'components.schemas.ParentFixture.properties') as unknown as Obj;
        const parentRequired = get(document, 'components.schemas.ParentFixture.required') as unknown[];
        expect(Object.keys(parentProperties)).toEqual(['foo', 'child']);
        expect(parentRequired).toEqual(['foo', 'child']);

        const childProperties = get(document, 'components.schemas.ChildFixture.properties') as unknown as Obj;
        const childRequired = get(document, 'components.schemas.ChildFixture.required') as unknown[];
        expect(Object.keys(childProperties)).toEqual(['baz']);
        expect(childRequired).toEqual(['baz']);

        expect(document).toMatchSnapshot();
    });
    it('redefines a field from a grand parent', async () => {
        class GrandParentFixture {
            @IsString()
            keep!: string;

            @IsString()
            redefine!: string;
        }

        const ParentFixture = pick(GrandParentFixture, 'Parent', ['keep'], ['redefine']);

        class ChildFixture extends ParentFixture {
            @IsInteger({
                optional: true,
            })
            redefine!: number;
        }

        const Controller = createController(ChildFixture);

        const document = await createDocument(Controller);
        const response = document.paths['/example']?.get?.responses[200];
        expect(response).toMatchObject({
            content: {
                'application/json': {
                    schema: {
                        $ref: '#/components/schemas/ChildFixture',
                    },
                },
            },
        });

        const childProperties = get(document, 'components.schemas.ChildFixture.properties') as unknown as Obj;
        const childRequired = get(document, 'components.schemas.ChildFixture.required') as unknown[];
        expect(Object.keys(childProperties)).toEqual(['keep', 'redefine']);
        expect(get(childProperties, 'redefine.type')).toEqual('integer');
        expect(childRequired).toEqual(['keep']);

        expect(document).toMatchSnapshot();
    });
});
