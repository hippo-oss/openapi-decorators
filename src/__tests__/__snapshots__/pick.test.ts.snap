// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pick picks a single field 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "ParentFixture": Object {
        "properties": Object {
          "foo": Object {
            "type": "string",
          },
        },
        "required": Array [
          "foo",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {},
    "description": "",
    "title": "Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "operationId": "ExampleController_find",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ParentFixture",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [],
  "tags": Array [],
}
`;

exports[`pick picks child fields 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Child": Object {
        "properties": Object {
          "baz": Object {
            "type": "string",
          },
          "qux": Object {
            "type": "string",
          },
        },
        "required": Array [
          "baz",
          "qux",
        ],
        "type": "object",
      },
      "ParentFixture": Object {
        "properties": Object {
          "child": Object {
            "$ref": "#/components/schemas/Child",
          },
        },
        "required": Array [
          "child",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {},
    "description": "",
    "title": "Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "operationId": "ExampleController_find",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ParentFixture",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [],
  "tags": Array [],
}
`;

exports[`pick picks one field and one child field 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "ChildFixture": Object {
        "properties": Object {
          "baz": Object {
            "type": "string",
          },
        },
        "required": Array [
          "baz",
        ],
        "type": "object",
      },
      "ParentFixture": Object {
        "properties": Object {
          "child": Object {
            "$ref": "#/components/schemas/ChildFixture",
          },
          "foo": Object {
            "type": "string",
          },
        },
        "required": Array [
          "foo",
          "child",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {},
    "description": "",
    "title": "Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "operationId": "ExampleController_find",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ParentFixture",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [],
  "tags": Array [],
}
`;

exports[`pick picks zero fields 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "ParentFixture": Object {
        "properties": Object {},
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {},
    "description": "",
    "title": "Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "operationId": "ExampleController_find",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ParentFixture",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [],
  "tags": Array [],
}
`;

exports[`pick redefines a field from a grand parent 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "ChildFixture": Object {
        "properties": Object {
          "keep": Object {
            "type": "string",
          },
          "redefine": Object {
            "type": "integer",
          },
        },
        "required": Array [
          "keep",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {},
    "description": "",
    "title": "Example",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/example": Object {
      "get": Object {
        "operationId": "ExampleController_find",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ChildFixture",
                },
              },
            },
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [],
  "tags": Array [],
}
`;
